### ISSUE TITLE
mockito-android has extremely large method count

### ISSUE THREAD

**Issue by drewhannay** [mockito/mockito#1112]
I've been investigating why my test apk has hit the 65k "multidex" method limit. It seems like `mockito-android` is by far the dominant contributor to method count.

To help get cleaner data, I created a brand new empty project in Android Studio 3.0-alpha3 and removed all dependencies except the following:
```
dependencies {
    androidTestImplementation 'org.mockito:mockito-android:2.8.9'
    androidTestImplementation('com.android.support.test.espresso:espresso-core:2.2.2', {
        exclude group: 'com.android.support', module: 'support-annotations'
    })
}
```

After this, I built the test apk using `./gradlew assembleDebugAndroidTest` and used the APK Analyzer tool in Android Studio to examine the APK. Here's what it reports:

`org.mockito` --> 2402 referenced methods
`net.bytebuddy` --> 13512 referenced methods
`com.android.dx` --> 4661 referenced methods

This totals to 20575 referenced methods, which is roughly 1/3 of the total allowable referenced method count for an Android test apk. Since test apks don't support legacy multidex, if a test apk goes over the 65k method limit, they can no longer run their tests on versions of android older than API 21 (Lollipop).

For extra info, here's the (relevant) output of `./gradlew :app:dependencies` for my sample project:

```
debugAndroidTestCompileClasspath - Resolved configuration for compilation for variant: debugAndroidTest
+--- org.mockito:mockito-android:2.8.9
|    +--- org.mockito:mockito-core:2.8.9
|    |    +--- net.bytebuddy:byte-buddy:1.6.14
|    |    +--- net.bytebuddy:byte-buddy-agent:1.6.14
|    |    \--- org.objenesis:objenesis:2.5
|    \--- net.bytebuddy:byte-buddy-android:1.6.5
|         +--- net.bytebuddy:byte-buddy:1.6.5 -> 1.6.14
|         \--- com.jakewharton.android.repackaged:dalvik-dx:1
|              \--- com.jakewharton.android.repackaged:libcore-dex:2
```

In my real app, we are already using Proguard on our test APK, and that only strips out ~3k of the 20k methods referenced by mockito-android (so this probably isn't a viable solution).

Is there anything we can do to reduce the method count pulled in by `mockito-android`?

**Comment by szczepiq** [comment]
Thank you for reporting!

@raphw, can you share your thoughts on reducing number of methods in bytebuddy? Is this something we can ask community to help out? Perhaps there are some cheap refactorings to do in bytebuddy to make proguard more successful with stripping out methods.

@drewhannay, I suspect that refactoring/thinning up bytebuddy is a major overhaul and unlikely to happen due to unknown ROI (we don't know the problem space, how many users are impacted and when proper solution from Google is available). I'd say our best bet at the moment is to tinker with Proguard, try to figure out how to squeeze more out of it, perhaps by changing/tweaking Mockito or Bytebuddy code.

**Comment by raphw** [comment]
@szczepiq I would like to try with the shade plugin. Also, the dexmaker mock maker might actually be an option for these cases as it does not implement any Java byte code specifc instrumentation. Byte Buddy would need to be excluded in this case.

**Comment by drewhannay** [comment]
Yes, some dexmaker users have reported success with excluding bytebuddy to reduce the method count (see https://github.com/linkedin/dexmaker/issues/63 )

**Comment by jaredsburrows** [comment]
@drewhannay Because of this, I continue to use:
```
  androidTestCompile "com.nhaarman:mockito-kotlin-kt1.1:1.5.0"      , { exclude group: "net.bytebuddy" } // DexMaker has it"s own MockMaker
  androidTestCompile "org.mockito:mockito-core:2.8.47"              , { exclude group: "net.bytebuddy" } // DexMaker has it"s own MockMaker
  androidTestCompile "com.linkedin.dexmaker:dexmaker-mockito:2.2.0" , { exclude group: "net.bytebuddy" } // DexMaker has it"s own MockMaker
```

**Comment by szczepiq** [comment]
>Yes, some dexmaker users have reported success with excluding bytebuddy to reduce the method count (see linkedin/dexmaker#63 )

I've looked at the ticket description, thank you for linking!

It seems that excluding "net.bytebuddy" is a reasonable workaround for the issue. I'm inclined to close the ticket and offer this as solution to the use case. Thoughts?

**Comment by drewhannay** [comment]
As long as this continues to work going forward, then it seems like a reasonable workaround to me.

**Comment by jaredsburrows** [comment]
@szczepiq I can try the following: 
```
  androidTestCompile "com.nhaarman:mockito-kotlin-kt1.1:1.5.0"      , { exclude group: "net.bytebuddy" } // DexMaker has it"s own MockMaker
  androidTestCompile "org.mockito:mockito-android:2.8.47"              , { exclude group: "net.bytebuddy" } // DexMaker has it"s own MockMaker
```

**Comment by liutikas** [comment]
Note, this this workaround is quite painful if you have 30+ modules in your project that all use mockito for testing, so I don't think we should close this unless there is no other way around this issue. For our projects, moving to mockito 2 and mockito-android caused us to move to multidex just for our test apks (no we do not want to run proguard for our test apks)

**Comment by tmurakami** [comment]
This workaround is effective only with DexMaker.
It does not work with `mockito-android` that [indirectly relies on Byte Buddy](https://github.com/mockito/mockito/blob/release/2.x/subprojects/android/src/main/java/org/mockito/android/internal/creation/AndroidByteBuddyMockMaker.java#L18).

**Comment by moltmann** [comment]
I am not quite sure about the long term relationship between mockito, mockito-android and dexmaker.

For now we have three mockmakers with different properties:

**mockito-android** using bytebuddy
\+ fully integrated with mockito
\- slowest mockmaker for Android
\- not an inline-mockmaker

**dexmaker-mockito**
\+ fastest known mockmaker for Android
\- not integrated
\- not an inline-mockmaker

**dexmaker-mockito-inline**
\+ inline-mockmaker
\- not integrated

I think the best integration in the long run would be if bytebuddy would natively support dex bytecode. As all this is not decided on yet, let me isolate the usage of bytebuddy in mockito-core so that we can do what @jaredsburrows suggests.

**Comment by mockitoguy** [comment]
Hey @moltmann thank you for feedback! What do you mean by "not integrated"?

**Comment by mockitoguy** [comment]
Guys, if mockito team shipped a new artifact "mockito-base" (say), that would not depend on any mockmaker (e.g. would not depend on bytebuddy), would that be helpful? (I'm not pitching an idea at this point, but rather I'm trying to understand the problem better... by offering a solution :).

**Comment by raphw** [comment]
I think that offering a core module is a good idea. Byte Buddy relies on the fact that most classes are never loaded what works for a standard VM but not for Android where all method are merged into one giant blob. Also, Byte Buddy aims to support the standard first and I do not think that I will work towards a native creation of dex classes, the library is complex enough as it is. Since Byte Buddy generates standard Java classes and then translates these class into the dex format, a generator that is tailored to Android will always be faster.

Personally, I still hope that standard Java classes might be supported by Android at some point as Google increasingly takes the OpenJDK into use. Limitations like the method count constraint were already resolved in more modern Android versions, in the long run the current limitations of Android might just disappear, but for the time being offering a core module might be the quickest path to success for those that require faster or inline tests on Android.

**Comment by moltmann** [comment]
@mockitoguy With "not integrated" I mean: Built, tested and developed inside of mockito. Hence all new features will work and there will no bugs

I looked into building such a plugin-free or at least mockmaker-free version of mockito and there are some issues:
- You need pull request 1320 to clean up the internal directory structure
- Currently default plugins are required to live inside of mockito. We enforce this via class loaders. Hence the -core module (which should pass the tests) cannot be the pluginfree version + the plugins but rather has to built as a new target. Maybe by gradle-fu isn't good enough, but I could not make this work nicely.
- As hinted above a plugin-free version of mockito would not pass the tests even if new plugins would be provided as having default plugins is enforced by the tests.

**Comment by mockitoguy** [comment]
Thank you for explaining!

Our long term strategy will focus on building strong API / platform and keeping non-core plugins external. We don't have sufficient Android expertise to confidently "own" the android integrations. We want to ensure that we have comprehensive test coverage and the features work 100% for all externally maintained plugins.

I merged the directory structure PR. Let's work on creating "mockito-base" module that would only rely on pure Java (not on bytebuddy).
